fix: select last used window when moving focus up or to the left

Daxtorim opened the pull request:
<!-- This won't be rendered!
[CHECKLIST]
- I read the contributing guide (CONTRIBUTING.md)
- My code follows the style guidelines of this project
- I have performed a self-review of my code
- I have commented on my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
-->

## Summary

When bismuth finds two windows are equally distanced from the current
window, the last used window will be chosen when focus is moved via
keyboard shortcuts.

Before, the closest windows were always determined based on the top-left
corner of each window. This is fine when focusing a window below or to
the right, since adjacent edge aligned windows will have the top-left
corner on the same line either horizontally or vertically.

When moving the focus up or to the left, however, this is no longer
guaranteed. Even though the bottom of two candidate windows might be
edge to edge with the current window, the top-left corners might be on
different levels since the candidate windows don't have to have the same
height. To correctly select the last used adjacent edge aligned window
to the top or to the left, the bottom-right corner will be used to find
the closest windows.

## Test Plan

1. Install PR version of script
2. Open 7 windows in the spiral layout
3. Go to the center most window
4. Go from there one window down (bottom center window)
5. Use Focus Upper Window shortcut
6. Observe how the previous window is focused instead of the shorter one to the right

---

What happens without this PR (the shorter window will *always* be chosen):

https://user-images.githubusercontent.com/72616153/143282973-7eaaed8b-865c-49d3-acc9-61fa5906197e.mp4

---

How it works now with this PR:

https://user-images.githubusercontent.com/72616153/143282824-fec0eea0-6850-4918-9f9b-5fb2b0be87ce.mp4

## Related Issues

#203 
This version still relies on the corners of each window to be perfectly pixel aligned to work properly. With gaps between windows which cause the corners to be offset from each other, even a single pixel, bismuth will always pick the closer window anyway.

gikari commented:
This doesn't work for me (I rebased this branch on master)

gikari commented:
Could you please test the behavior of directional switching using the native KWin Shortcuts for this? Is there the same "chose the most recently used window between two equally distant ones" behavior? There are plans to remove the Bismuth shortcuts for directional switching in the 3.0 version and replace them with native KWin ones.

Daxtorim commented:
> This doesn't work for me (I rebased this branch on master)

Hmm, i rebased it locally as well and it *does* work for me. Are you sure you aren't seeing #203? On a high dpi screen that would be hard to see. The windows really need to be pixel perfectly aligned for this to work. I suppose I could give it a couple pixels wiggle room.

> Could you please test the behavior of directional switching using the native KWin Shortcuts for this? 

Frankly, that implementation is not acceptable. 

- You cannot focus a window if there was none focused in the first place. Similar to the issue that was fixed in #208, except it doesn't matter how many windows are on the screen, you just can't focus a window without the mouse (or `Alt + Tab`, but eh).
- Sometimes when focusing the next window left or right, it will instead jump up or down, making for a really jarring experience.
- There is some attempt to wrap around, i.e. go to the bottom most window when going off the top of the screen, but it is buggy at best, jumping to a middle window instead if not just to the right or left. 

In the spiral layout, the native shortcuts are a mess to use and seemingly jump around randomly at times. Based on my couple of minutes of testing them, I don't think those are a good replacement without some work upstream.

> Is there the same "chose the most recently used window between two equally distant ones" behavior?

No. Based on the above, I don't know if there is any special logic to choosing which window get focus.

Daxtorim commented:
I added a few pixel leeway to the part that determines the closest windows. That should avoid issues with misaligned tiles. 

I don't really understand the code that determines the size of each tile, so I can't accurately say what the error margin is that causes #203. That's why I chose the 5 pixel boundary, it should avoid the issue without being *too* lenient.

gikari commented:
Now it works.

gikari commented on src/kwinscript/engine/index.ts:
```diff
@@ -630,22 +630,27 @@ export class EngineImpl implements Engine {
     dir: Direction
   ): EngineWindow | null {
     let vertical: boolean;
+    let downOrRight: boolean;
```
It's not clear what this variable is about. It's also true about `vertical`. Could you rename it so that is not needed to read the following lines to understand what it represents?

For example, `directionIsVertical` and `directionIsDownOrRight`?

I am also not sure if it's good to place all this flags in the switch case, while it could be achieved via simple:
```typescript
    const downOrRight = dir === "down" || dir === "right";
```

The same is for the other flags.

gikari commented on src/kwinscript/engine/index.ts:
```diff
@@ -680,21 +685,34 @@ export class EngineImpl implements Engine {
 
+    // When moving downOrRight, check the top-left corner for closest window
+    // When moving not downOrRight, check the bottom-right corner for closest window
     const min =
```
1. Bad variable name.
2. This var value acquisition logic should be extracted to a separate function.

gikari commented on src/kwinscript/engine/index.ts:
```diff
@@ -680,21 +685,34 @@ export class EngineImpl implements Engine {
+        }
+      }, Infinity);
+
+    const closest = candidates.filter((tile): boolean => {
```
This var value acquisition logic should be extracted to a separate function.

gikari commented on src/kwinscript/engine/index.ts:
```diff
@@ -680,21 +685,34 @@ export class EngineImpl implements Engine {
+      }
+    });
 
     return closest.sort((a, b) => b.timestamp - a.timestamp)[0];
```
To make this a little more obvious, there should be a comment, that sounds something like this:
```js
// Get the most recently used window
```

gikari commented on src/kwinscript/engine/index.ts:
```diff
@@ -630,22 +630,27 @@ export class EngineImpl implements Engine {
+        downOrRight = true;
         sign = 1;
         break;
       default:
```
GitHub does not let me add a comment to the correct line, but the candidates var should be extracted to a separate function.

gikari reviewed (changes requested):
Since you're adding more code, could you please refactor the `getNeghborByDirection` function (taking my comments into account)?

Daxtorim commented:
I changed the variable names and replaced the switch statement as you suggested.

---

> This var value acquisition logic should be extracted to a separate function.

I'm not entirely sure what you mean. I guess something like 
```min = dirSign * candidates.reduce(findClosestWindow, Infinity)```
instead of the inline function definition. The issue is that these functions need the flags that were set above and thus would need them as parameters. But `filter()` and `reduce()` only pass predefined parameters to the callback function.

I tried to wrap the callback like this:
```
private findClosestWindow(prevMin, window, dirIsVertical, dirIsDownOrRight, dirSign): number {
  [...]
}

[...]

min = dirSign * candidates.reduce(
  (prevMin, tile) => {
    return this.findClosestWindow(
      prevMin, tile, dirIsVertical, dirIsDownOrRight, dirSign
    )
  }
)
```
but I couldn't get this working.

gikari commented:
> I'm not entirely sure what you mean. 

You need to extract the long reduce/filter function with nested `if`s to a private member of the `Engine` class. You don't need to pass the flags (they are primitive), just pass the direction and the basis, and return the result of the filter or reduce.

For example:

```typescript
    const candidates = this.windows
      .visibleTiledWindowsOn(this.controller.currentSurface)
      .filter(
        vertical
          ? (tile): boolean => tile.geometry.y * sign > basis.geometry.y * sign
          : (tile): boolean => tile.geometry.x * sign > basis.geometry.x * sign
      )
      .filter(
        vertical
          ? (tile): boolean =>
              overlap(
                basis.geometry.x,
                basis.geometry.maxX,
                tile.geometry.x,
                tile.geometry.maxX
              )
          : (tile): boolean =>
              overlap(
                basis.geometry.y,
                basis.geometry.maxY,
                tile.geometry.y,
                tile.geometry.maxY
              )
      );

```

becomes:

```typescript
      const neighborCandidates = this.getNeighborCandidates(dir);
```
and
```typescript
  private getNeighborCandidates(
    basis: EngineWindow,
    dir: Direction
  ): EngineWindow[] {
    const visibleWindowsOnCurrentSurface = this.windows.visibleTiledWindowsOn(
      this.controller.currentSurface
    );

    // Flipping all inputs' signs allows for the same logic to find closest windows in either direction
    const sign = dir === "down" || dir === "right" ? 1 : -1;

    if (dir === "up" || dir === "down") {
      return visibleWindowsOnCurrentSurface.filter(
        (tile): boolean =>
          tile.geometry.y * sign > basis.geometry.y * sign &&
          overlap(
            basis.geometry.x,
            basis.geometry.maxX,
            tile.geometry.x,
            tile.geometry.maxX
          )
      );
    } else {
      return visibleWindowsOnCurrentSurface.filter(
        (tile): boolean =>
          tile.geometry.x * sign > basis.geometry.x * sign &&
          overlap(
            basis.geometry.y,
            basis.geometry.maxY,
            tile.geometry.y,
            tile.geometry.maxY
          )
      );
    }
  }

```

It may seem to be verbose, but when you use the above refactor on three functions, the body of the `getNeighborByDirection` would consist of ~3 function calls and its logic would be more clear, so that the reader would not hate to try to dig into implementation details of these parts.
